dsl/flatten.cc

```diff
@@ -0,0 +1,216 @@
+    // DSL pass.
+    bool isMethodModifier(ast::Send &send) {
+        auto fun = send.fun;
+        return (fun == core::Names::private_() || fun == core::Names::protected_() || fun == core::Names::public_() ||
+                fun == core::Names::privateClassMethod()) &&
+               send.args.size() == 1 && ast::isa_tree<ast::MethodDef>(send.args[0].get());
+    }
+
+    // We might want to move sends as well: either if they're method modifiers like `private` or `protected` or if
+    // they're `sig`s. If so, then we'll treat them like we treat methods on our method stack
+    unique_ptr<ast::Send> preTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        if (send->fun == core::Names::sig() || isMethodModifier(*send)) {
+            auto &methods = curMethodSet();
+            bool isStatic = methods.stack.size() > 0 && methods.stack.back().isStatic;
+            methods.stack.emplace_back(methods.methods.size(), isStatic);
+            methods.methods.emplace_back();
+
+            if (isMethodModifier(*send) && send->args.size() >= 1) {
+                // if this is a method modifier like `private` or `protected`, then we don't need to add a new scope
+                // when we traverse the method itself, so add it to the ignore set...
+                skipMethods.insert(send->args[0].get());
+                // but we want /this/ stack frame to be a static frame if the method def it wraps is, so we should
+                // update our stack frame metadata accordingly
+                if (auto methodDef = ast::cast_tree<ast::MethodDef>(send->args[0].get())) {
+                    methods.stack.back().isStatic |= methodDef->isSelf();
+                }
+            }
+        }
+        return send;
+    }
+
+    unique_ptr<ast::Expression> postTransformSend(core::Context ctx, unique_ptr<ast::Send> send) {
+        if (send->fun == core::Names::sig() || isMethodModifier(*send)) {
+            auto &methods = curMethodSet();
+            ENFORCE(!methods.stack.empty());
+            ENFORCE(methods.methods.size() > methods.stack.back().idx);
+            ENFORCE(methods.methods[methods.stack.back().idx] == nullptr);
+
+            methods.methods[methods.stack.back().idx] = std::move(send);
+            methods.stack.pop_back();
+
+            return make_unique<ast::EmptyTree>();
+        } else {
+            return send;
+        }
+    }
+
+    unique_ptr<ast::Expression> postTransformClassDef(core::Context ctx, unique_ptr<ast::ClassDef> classDef) {
+        classDef->rhs = addMethods(ctx, std::move(classDef->rhs));
+        return classDef;
+    };
+
+    unique_ptr<ast::Expression> postTransformMethodDef(core::Context ctx, unique_ptr<ast::MethodDef> methodDef) {
+        // if this method is contained in a send like `private` or `protected`, then we should not move it, because
+        // moving the send will do that for us
+        if (skipMethods.contains(methodDef.get())) {
```

Comment:
lets address https://github.com/sorbet/sorbet/pull/2009/files#r338308368 first, because this for me seems ad-hoc, but I might not have context.